#Generating Game and Car Dataset

In [ ]:
!pip install newspaper3k

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.1/211.1 kB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.1/81.1 kB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.7/97.7 kB 12.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 78.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Created wheel for tinysegmenter: filename=tinysegmenter-0.3-py3-none-any.whl size=13538 sha256=4d098448db59dcb772a1c21afd49a3c4d82d810108aab164c3aab2e51ac9fc49
  Stored in directory: /root/.cache/pip/wheels/c8/d6/6c/384f58df48c00b9a31d638005143b5b3ac62c3d25fb1447f23
  Created wheel for feedfinder2: filename=feedfinder2-0.0.4-py3-none-any.whl size=3340 sha256=46528ab61b2a5bb48c6fb9e5e0c8a670ca396a45895972deea9e07dc79ed981c
  Stored in directory: /root/.cache/pip/wheels/97/02/e7/a1ff1760e12bdbaab0ac824fae5c1b

#Scraping all Game related articles from various websites and assigning the label "Game"

We created our custom dataset by scraping data from open websites (gamespot.com, ign.com, topgear.com, and autoblog.com) using the newspaper3k library, as listed below. We ran these cells once to create a csv file, which we will use for further processing.

In [ ]:
import sqlite3
import newspaper
import random
import csv

conn = sqlite3.connect('game_articles.db')
c = conn.cursor()

c.execute('CREATE TABLE IF NOT EXISTS articles (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, text TEXT, label TEXT)')


websites = ['https://store.steampowered.com/', 'https://www.ign.com/',
            'https://www.gamespot.com/', 'https://www.theverge.com/games',
            'https://en.wikipedia.org/wiki/Game']

for website in websites:
    print(f'Scraping {website}...')
    paper = newspaper.build(website)
    count = 0
    for article in paper.articles:
        if count == 200:
            break
        try:
            article.download()
            article.parse()
            title = article.title
            text = article.text
            c.execute("INSERT INTO articles (title, text, label) VALUES (?, ?, ?)", (title, text, 'game'))
            count += 1
        except:
            pass

conn.commit()

c.execute("SELECT title, text, label FROM articles WHERE label='game'")
results = c.fetchall()

random.shuffle(results)

with open('game_dataset.csv', 'w', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(['text', 'label'])
    for result in results:
        writer.writerow([result[1], 'game'])

conn.close()


Scraping https://store.steampowered.com/...
Scraping https://www.ign.com/...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/esports
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/NONE
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/entertainment
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/gamescom-asia
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.ign.com/rss


Scraping https://www.gamespot.com/...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.gamespot.com/rss/
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.gamespot.com/feed/


Scraping https://www.theverge.com/games...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.theverge.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.theverge.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.theverge.com/rss


Scraping https://en.wikipedia.org/wiki/Game...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://en.wikipedia.org/rss
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://en.wikipedia.org/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://en.wikipedia.org/feed


# Scraping all car related articles from various websites and assigning the label "car"

In [ ]:
import sqlite3
import newspaper
import random
import csv

conn = sqlite3.connect('car_articles.db')
c = conn.cursor()

c.execute('CREATE TABLE IF NOT EXISTS articles (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, text TEXT, label TEXT)')

websites = ['https://www.caranddriver.com/news/', 'https://www.motortrend.com/', 'https://www.autoblog.com/',
            'https://www.topgear.com/car-news', 'https://www.cnn.com/business/cars']

for website in websites:
    print(f'Scraping {website}...')
    paper = newspaper.build(website)
    count = 0
    for article in paper.articles:
        if count == 200:
            break
        try:
            article.download()
            article.parse()
            title = article.title
            text = article.text
            c.execute("INSERT INTO articles (title, text, label) VALUES (?, ?, ?)", (title, text, 'car'))
            count += 1
        except:
            pass

conn.commit()
c.execute("SELECT title, text, label FROM articles WHERE label='car'")
results = c.fetchall()

random.shuffle(results)

with open('car_dataset.csv', 'w', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(['text', 'label'])
    for result in results:
        writer.writerow([result[1], 'car'])

conn.close()

Scraping https://www.caranddriver.com/news/...


CRITICAL:newspaper.network:[REQUEST FAILED] 403 Client Error: Forbidden for url: https://www.caranddriver.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 403 Client Error: Forbidden for url: https://www.caranddriver.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 403 Client Error: Forbidden for url: https://www.caranddriver.com/rss


Scraping https://www.motortrend.com/...
Scraping https://www.autoblog.com/...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://guce.autoblog.com/
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error:  for url: https://www.autoblog.com/feed/
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error:  for url: https://www.autoblog.com/rss/


Scraping https://www.topgear.com/car-news...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.topgear.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.topgear.com/feeds
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://www.topgear.com/rss


Scraping https://www.cnn.com/business/cars...


CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://edition.cnn.com/feed
CRITICAL:newspaper.network:[REQUEST FAILED] 404 Client Error: Not Found for url: https://edition.cnn.com/feeds


#Importing and exploring the custom datasets.

In [ ]:
import pandas as pd
import numpy as np

game_df = pd.read_csv('game_dataset.csv')
car_df = pd.read_csv('car_dataset.csv')

In [ ]:
game_df

text label
0    Comparison Grid\n\nComplete edition\n\n\n\nInc...  game
1    Disney Dreamlight Valley's A Rift in Time expa...  game
2    Low-profile mechanical keyboards have always s...  game
3    Goodbye, 13-inch MacBook Pro. I will not miss ...  game
4    Unsurprisingly, Black Friday and Cyber Monday ...  game
..                                                 ...   ...
761  Marie-Joseph Paul Yves Roch Gilbert du Motier ...  game
762  Quy trình đề cử\n\nHãy xem các tiêu chuẩn của ...  game
763  The Season of the Wish has dredged up some old...  game
764  Your Warhammer 40K: Rogue Trader homeworld isn...  game
765  These days, we are all dealing with huge amoun...  game

[766 rows x 2 columns]

In [ ]:
car_df

text label
0    Like our planet's endangered climate, this lis...   car
1    Top Gear\n\nNewsletter\n\nThank you for subscr...   car
2    لم يعد ما يُسمى بـ "الموضة السريعة" رائجًا فيم...   car
3    WASHINGTON — Billionaire businessman Elon Musk...   car
4    STOCKHOLM, Sweden — Tesla lost a chapter in a ...   car
..                                                 ...   ...
493  Mitchell Layton/Getty Images\n\nThe NCAA is re...   car
494  El primer Alfa Romeo eléctrico está muy cerca ...   car
495  ¿Qué es exactamente?\n\nEl Plan Moves III es u...   car
496  The all-new LBX is the smallest Lexus ever\n\n...   car
497  Sign up for CNN’s Wonder Theory science newsle...   car

[498 rows x 2 columns]

In [ ]:
df_combined = pd.concat([car_df, game_df], ignore_index=True)
df_combined = df_combined.sample(frac=1, random_state=23).reset_index(drop=True)
df_combined.to_csv('combined_dataset.csv', index=False)

In [ ]:
df_combined.head()

text label
0  Brussels/London CNN —\n\nThe European Union is...   car
1  The 2022 F-150 has to be many things to many p...   car
2  If you've ever wanted to start a bit of drama,...  game
3  As part of our cover story on director Zack Sn...  game
4  It’s here. Everybody be cool, but the reveal t...   car

In [ ]:
df_combined.shape

(1264, 2)

As we see that the total number of rows is around 1264, it is CLEARLY evident that
there are at least more than 1200+ words assuming each sentence will surely at least
have 1 word (a lot more!!).

#Keeping only English articles

In [ ]:
!pip install langdetect

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 13.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993225 sha256=975c8bef2908dc9c9031a9ee2de52e3139281a7c0b1f0882de847d5f2ebd3549
  Stored in directory: /root/.cache/pip/wheels/95/03/7d/59ea870c70ce4e5a370638b5462a7711ab78fba2f655d05106
Successfully built langdetect


In [ ]:
import langdetect

def detect_language(text):
    try:
        return langdetect.detect(text)
    except:
        return 'unknown'

df_combined['language'] = df_combined['text'].apply(detect_language)

In [ ]:
df_combined.label.value_counts()

game    766
car     498
Name: label, dtype: int64

In [ ]:
df_combined = df_combined[df_combined['language'] == 'en'].iloc[:,:2]

Randomly choose two sentences from each article to reduce the data

In [ ]:
import random

def get2sentences(row):
    if row == "":
        return row
    row = row.replace('\n',".")
    row = row.replace('!',".")
    sentences = row.split(".")
    random.shuffle(sentences)
    if len(sentences)>4:
        return ".".join(sentences[:2])+"."
    return row

In [ ]:
df_combined = df_combined.reset_index(drop=True)
df_combined['text'] = df_combined['text'].apply(get2sentences)

In [ ]:
df_combined.shape

(986, 2)

#Splitting the data into train and test samples

We have split the dataset into train‑test ratio of 0.2 which means there are at least 190+ samples of
testing and 780+ training samples


In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_combined.text, df_combined.label, test_size=0.2, random_state=23)

In [ ]:
X_train.head()

146    .7 Verge Score Lofree Flow $ 152 The Good Slee...
408                                                   ..
398     But just given everything that was going on, ...
268    .“With the Neue Klasse, we have embarked on th...
307    Speaking to The Hollywood Reporter, Cage said ...
Name: text, dtype: object

#Tokenize the data using the AutoTokenizer from "bert‑base‑cased" model.

Since the BERT-base-cased model has been trained on a vast and varied corpus of text data, which includes language related to both games and cars, it is effective at predicting labels for sentences pertaining to the comparison of games and cars. The model's comprehension of pertinent vocabulary and semantic structures can be utilized to make precise predictions when it is fine-tuned for a task of classifying sentences related to games versus cars.

In [ ]:
!pip install transformers

In [ ]:
from transformers import AutoTokenizer
from transformers import TFAutoModelForSequenceClassification
from tensorflow.keras.optimizers import Adam
import numpy

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
def tokenize(X):
    #truncation for not exceeding 512 limit
    tokenized_data = tokenizer(X, return_tensors="np", truncation=True, padding=True)
    tokenized_data = dict(tokenized_data)
    return tokenized_data

tokenizer_config.json:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

In [ ]:
X_train = tokenize(list(X_train.astype(str)))
X_test = tokenize(list(X_test.astype(str)))

from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()
y_train= label_encoder.fit_transform(y_train)

label_encoder = preprocessing.LabelEncoder()
y_test= label_encoder.fit_transform(y_test)

In [ ]:
model = TFAutoModelForSequenceClassification.from_pretrained("bert-base-cased")
model.compile(optimizer=Adam(3e-5))
model.fit(X_train, y_train)

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFBertForSequenceClassification.

Some weights or buffers of the TF 2.0 model TFBertForSequenceClassification were not initialized from the PyTorch model and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


25/25 [==============================] - 74s 798ms/step - loss: 0.6317


In [ ]:
y_pr = model.predict(X_test)["logits"]
y_pred = np.argmax(y_pr, axis=1)

from sklearn.metrics import accuracy_score, confusion_matrix
confusion_matrix(y_true=y_test, y_pred=y_pred)
print('Accuracy: %.3f' % accuracy_score(y_test, y_pred))

7/7 [==============================] - 4s 213ms/step
Accuracy: 0.813


We obtain a test accuracy of 81.3% which is pretty good for random data scraped from the internet which was a part of articles related to the 2 labels, car and game.

#Ways to improve Accuracy.

We can adjust the model to the specific data by taking into account the context of the incoming data, which will further IMPROVE accuracy. Since our goal is to predict car specifications labels, for instance, training on data from technical specification articles or review magazines will be far more beneficial than simply scraping random car sales data or pollution-related data from a specific dealership. Many keywords that are closely related will be introduced by these.

To improve the model's training, we can also expand the amount of training data.

To determine which pre-trained model performs best when data changes, we can experiment with various models.